File: insert_sort/insert_sort.c
#include <stdio.h>
#define MAXSIZE 20

// typedef int KeyType;

// typedef struct
// {
//     KeyType key;
// } RedType;

// typedef struct
// {
//     RedType r[MAXSIZE + 1];
//     int length;
// } SqList;

void InsertSort(int *arr, int len)
{
    int temp, i, j;
    for (i = 1; i < len; i++)
    {
        if (arr[i - 1] > arr[i])
        {
            temp = arr[i];
            for (j = i - 1; j >= 0 && arr[j] > temp; j--)
            {
                arr[j + 1] = arr[j];
            }
            arr[j + 1] = temp;
        }
    }
}

int main()
{
    int arr[] = {2, 5, 3, 5, 6, 8, 1};
    InsertSort(arr, 7);
    for (int i = 0; i < 7; i++)
    {
        printf("%d ", arr[i]);
    }
    printf("\n");
    return 0;
}

File: insert_sort/insert.sort.go
package main

import "fmt"

func InserSort(arr []int) {
	var i, j, temp int
	for i = 1; i < len(arr); i++ {
		if arr[i-1] > arr[i] {
			temp = arr[i]
			for j = i - 1; j >= 0 && arr[j] > temp; j-- {
				arr[j+1] = arr[j]
			}
			arr[j+1] = temp
		}
	}
}

func main() {
	arr := []int{2, 5, 3, 5, 6, 8, 1}
	fmt.Println(len(arr))
	InserSort(arr)
	fmt.Println(arr)
}
